### PR TITLE
Important refactoring of the code

### PR DESCRIPTION
**Pull request by dpiras**
We essentially changed the syntax to call GMM-MI. The `fit` method now fits the GMM to the data only, and MI is estimated with `predict`. One can directly fit and predict by calling `fit_predict`, just like in `sklearn`.

Other additions include:
* can also compute KL between marginals, both `forward` and `reverse` (KL is not symmetric). Simply indicate `kl=True, kl_order='forward'` in the `predict` or `fit_predict` methods. Note this will compute both MI and KL; if `kl=False`, only MI will be computed. Use `kl_order='inverse'` to compute the inverse KL. KL is accessed with `mi_estimator.KL_mean` and  `mi_estimator.KL_std`.
* Can change units of MI and KL by specifying `base` in the `predict` or `fit_predict` methods (this is the base of the logarithm).
* All metrics (validation score, AIC and BIC) are stored during fitting, and can be accessed by e.g. 
```
for n_c in mi_estimator.results_dict.keys():
    # access AIC for all components fitted, even if another metric was chosen
    print(mi_estimator.results_dict[n_c]['aic']) 
```
* Can also pass 1D arrays as input, instead of stacking them (as discussed in https://github.com/dpiras/GMM-MI/issues/4)
* All these changes are included and explained in the walkthrough notebook and README.
* All notebooks have been adapted to reflect these changes.

Once we merge this PR, we'll also make this new version of GMM-MI `pip` installable.

### CODE REVIEW

**Comment by dpiras** [comment]
@lluciesmith I worked to refactor the code to include all the changes we discussed. If you have time, let me know if you agree with the changes, if you think other features should be added, and if you are able to use GMM-MI as intended. 

Then I'll make this version `pip` installable.

No rush of course. Thank you!

**Comment by lluciesmith** [comment]
@dpiras thank you for all the changes. I am going through them but it may take a while since a lot has changed.
One first comment: I found it a bit confusing that `mi_estimator.predict(kl=True)` does not actually return the KL (I thought it was a bug at first). Is there a reason why you decided to implement it this way? I very much like the fact that you can always access the metrics via `mi_estimator.MI_mean` or `mi_estimator.KL_mean`, but I would have also expected something like this:

```
mi_estimator = EstimateMI()
mi_estimator.fit(a,b)
MI_mean, MI_std = mi_estimator.predict()
KL_mean, KL_std = mi_estimator.predict(kl=True)
```

PS: I feel like `predict` doesn't quite work in this context but I can't really think of anything better. Maybe `estimate`/ `fit_estimate`?

**Comment by dpiras** [comment]
Thank you, @lluciesmith! It might be easier for you to check that the estimator works as intended in your workflow, so that you don't have to go through hundreds of line changes (which is essentially just the same code as before, just organised differently) 😅

The reason for not returning the KL directly is that I thought `kl=True` should return _also_ the KL (together with MI); in other words, I think that GMM-MI users will almost always require MI first, and then sometimes _also_ KL. The only problem I see with the current setting is computational cost: if the user wants _only_ KL, then MI calculation is only slowing everything down. I'm not sure it would be useful to provide the option to _only_ compute KL, since I don't know how often that would be used, but maybe you have a better sense of whether it'd be useful or not? It would be an easy fix in case.

I agree `estimate` is a better choice than `predict`, which I chose so far only because `fit`/`predict` are common in `sklearn`, which is quite popolar. Happy to change to `estimate` though if you also feel it's better!

**Comment by lluciesmith** [comment]
@dpiras yep I am trying the code out rather than checking all the lines in the code.

Intuitively, I would think that if the user manually sets `kl=True` it is probably because they want to compute KL instead of MI in that instance. Here are two options: (1) keep `kl=True` argument and return the KL when that is set, (2) change the argument to `add_kl=True` or `include_kl=True` and always return MI but also store the KL in the class (as you have currently implemented). I think the arguments are clearer in this way, what do you think? Both options are equally fine to me so you can make that choice -- indeed your point about computational cost may mean that (1) is more efficient but as you said most people will people user GMMMI for MI anyway?

**Comment by lluciesmith** [comment]
The fact that one can access different metrics to assess the fits is a great addition. However, I am wondering if these should all be computed automatically (as done currently) or if the user should be given the ability to choose which additional metrics to store? This could be easily done by having an additional argument called 'metrics' in the GMM param holder. I suppose the question is how computationally expensive is it to compute these metrics every time? If it is completely negligible then I suppose there is no harm in storing them all by default.

**Comment by dpiras** [comment]
Thank you, @lluciesmith!

> Here are two options: (1) keep kl=True argument and return the KL when that is set, (2) change the argument to add_kl=True or include_kl=True and always return MI but also store the KL in the class (as you have currently implemented). I think the arguments are clearer in this way, what do you think? Both options are equally fine to me so you can make that choice -- indeed your point about computational cost may mean that (1) is more efficient but as you said most people will people user GMMMI for MI anyway?

I would go with option number 2 for now, i.e. change the argument to `include_kl`. If we see that users need to estimate KL more than MI, and computing KL together with MI slows things down, we could add a boolean flag `compute_MI` in the `estimate` method that gives further flexibility to run only the KL and not the MI estimation, or indeed implement option 1 that you described.

> I suppose the question is how computationally expensive is it to compute these metrics every time? If it is completely negligible then I suppose there is no harm in storing them all by default.

I think the extra cost is negligible; computing `aic` or `bic` basically requires a single extra fit (as in [here](https://github.com/dpiras/GMM-MI/blob/refactoring/gmm_mi/mi.py#L170)) with the initial parameters already close to the optimum. Also, if one uses the validation score as metric and additionally requires `aic`, `bic` comes essentially for free; in one uses `aic` as the metric, then `bic` or `valid_score` come for free too. In some little experimenting I did I saw no important changes in the runtime.

Once you give me the thumbs up, I will:
- change `kl` to `include_kl`;
- change `predict` (`fit_predict`) to `estimate` (`fit_estimate`)
- anything else, just let me know!

**Comment by lluciesmith** [comment]
Great! I am happy with all the changes. Thanks a lot Davide

**Comment by dpiras** [comment]
I updated all of the above, including the `sklearn` problem of <=1.1 vs >= 1.2, where the `GaussianMixture` method [`_check_initial_parameters`](https://github.com/scikit-learn/scikit-learn/blob/1.1.X/sklearn/mixture/_base.py#L237) was changed to [`_check_parameters`](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/mixture/_base.py#L225). Will run some tests locally just to double check everything works as intended. This new version will be released as `0.4.1`.

Thank you so much for your contribution @lluciesmith! Let me know if there are other fixes/improvements to make 💪
